src/lib/chawathe98_matcher.rs

```diff
@@ -46,35 +46,2262 @@
+
+        let remaining_edges = edges_still_left.clone();
+
+        for e1_root in remaining_edges.clone() {
+            // Check if its root
+            if e1_root.from_node.is_root(&input.from_arena.borrow()) {
+                // If it's root it should not have move
+                // It could have glue, copy
+                if e1_root.edge_type != EdgeType::MOVE {
+                    remaining_edges_remove_root_move.push(e1_root.clone());
+                }
+            } else if e1_root.to_node.is_root(&input.to_arena.borrow()) {
+                // To node should have anything mapped to the root
+                // Same logic as the if statement above
+                if e1_root.edge_type != EdgeType::MOVE {
+                    remaining_edges_remove_root_move.push(e1_root.clone());
+                }
+            } else {
+                remaining_edges_remove_root_move.push(e1_root.clone());
+            }
+        }
+
+        edges_still_left = remaining_edges_remove_root_move;
+
+        // Remove weird move operation and copy operation
+        // If move and copy exist exist together take the one which has the same descendants
+        // And Both Map to the same node
+        let all_edges_remaining = edges_still_left.clone();
+        let mut remaining_edges: Vec<Edge> = Vec::new();
+        for edge_move in all_edges_remaining.clone() {
+            let m_move = edge_move.from_node;
+            let n_move = edge_move.to_node;
+            let mut check = false;
+            if edge_move.edge_type == EdgeType::MOVE {
+                for edge_glue in all_edges_remaining.clone() {
+                    let m_glue = edge_glue.from_node;
+                    let n_glue = edge_glue.to_node;
+
+                    if edge_glue.edge_type == EdgeType::GLUE
+                        && edge_move.edge_type == EdgeType::MOVE
+                        && m_move == m_glue && n_move == n_glue
+                        {
+                            // We have found a copy so check is true
+                            // so we have found (m,n) which is move and copy
+                            check = true;
+                            // We add copy edge to the remaining edges
+                            remaining_edges.push(edge_glue.clone());
+                        }
+                }
+            } else if edge_move.edge_type != EdgeType::GLUE {
+                remaining_edges.push(edge_move.clone());
+            }
+            if check == false && edge_move.edge_type == EdgeType::MOVE {
+                remaining_edges.push(edge_move.clone());
+            }
+        }
+        edges_still_left = remaining_edges;
+
+        /*
+        Questions.
+                 1) Do we need to consider where when there multiple (m,n) edge ?
+        */
+
+        return (edges_still_left, edges_to_be_pruned);
+    }
+}
+
+/// Creates a simple tree for tests which is FromNodeId
+pub fn create_arena() -> Arena<String, FromNodeId> {
```

Comment:
I think this code is used in your tests and not in the core algorithms, so it needs to be moved into the test module below that starts: `mod tests`.